lang/ir/src/ir/item/event.rs

```diff
@@ -0,0 +1,386 @@
+// limitations under the License.
+
+use crate::{
+    error::ExtError as _,
+    ir,
+};
+use core::convert::TryFrom;
+use proc_macro2::{
+    Ident,
+    Span,
+};
+use syn::spanned::Spanned as _;
+
+/// An ink! event struct definition.
+///
+/// # Example
+///
+/// ```
+/// # use core::convert::TryFrom;
+/// # let event = <ink_lang_ir::Event as TryFrom<syn::ItemStruct>>::try_from(syn::parse_quote! {
+/// #[ink(event)]
+/// pub struct Transaction {
+///     #[ink(topic)]
+///     from: AccountId,
+///     #[ink(topic)]
+///     to: AccountId,
+///     value: Balance,
+/// }
+/// # }).unwrap();
+/// ```
+#[derive(Debug, PartialEq, Eq)]
+pub struct Event {
+    item: syn::ItemStruct,
+}
+
+impl quote::ToTokens for Event {
+    /// We mainly implement this trait for this ink! type to have a derived
+    /// [`Spanned`](`syn::spanned::Spanned`) implementation for it.
+    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+        self.item.to_tokens(tokens)
+    }
+}
+
+impl Event {
+    /// Returns `true` if the first ink! annotation on the given struct is
+    /// `#[ink(event)]`.
+    ///
+    /// # Errors
+    ///
+    /// If the first found ink! attribute is malformed.
+    pub(super) fn is_ink_event(
+        item_struct: &syn::ItemStruct,
+    ) -> Result<bool, syn::Error> {
+        if !ir::contains_ink_attributes(&item_struct.attrs) {
+            return Ok(false)
+        }
+        // At this point we know that there must be at least one ink!
+        // attribute. This can be either the ink! storage struct,
+        // an ink! event or an invalid ink! attribute.
+        let attr = ir::first_ink_attribute(&item_struct.attrs)?
+            .expect("missing expected ink! attribute for struct");
+        Ok(matches!(attr.first().kind(), ir::AttributeArgKind::Event))
+    }
+}
+
+impl TryFrom<syn::ItemStruct> for Event {
+    type Error = syn::Error;
+
+    fn try_from(item_struct: syn::ItemStruct) -> Result<Self, Self::Error> {
+        let struct_span = item_struct.span();
+        let (_ink_attrs, other_attrs) = ir::sanitize_attributes(
+            struct_span,
+            item_struct.attrs,
+            &ir::AttributeArgKind::Event,
+            |kind| !matches!(kind, ir::AttributeArgKind::Event),
+        )?;
+        if !item_struct.generics.params.is_empty() {
+            return Err(format_err_spanned!(
+                item_struct.generics.params,
+                "generic ink! event structs are not supported",
+            ))
+        }
+        match &item_struct.vis {
+            syn::Visibility::Inherited
+            | syn::Visibility::Restricted(_)
+            | syn::Visibility::Crate(_) => {
+                return Err(format_err_spanned!(
+                    &item_struct.vis,
+                    "non `pub` ink! event structs are not supported",
+                ))
+            }
+            _ => (),
+        }
+        'outer: for field in item_struct.fields.iter() {
```

Comment:
TIL, I thought it was a lifetime. Though as I understand this is for nested loops, where is the `'inner`?

Reply:
Yeah I use it always when having to use `continue` or `break` (goto) statements because to me the code is more readable and more robust in case a person changes the code and might introduce new loops etc.

Reply:
Maybe it is bad style so it is more of a personal thing. If there are good reasons to avoid those annotations in general comments are very welcome!

Reply:
All good I just had never seen it before - as in the other comment I agree it is good for readability and robustness